shared/src/pool_cache.rs

```diff
@@ -0,0 +1,521 @@
+    /// block can have to be considered.
+    pub fn new(
+        number_of_blocks_to_cache: NonZeroU64,
+        number_of_pairs_to_auto_update: usize,
+        maximum_recent_block_age: u64,
+        inner: Box<dyn PoolFetching>,
+        block_stream: CurrentBlockStream,
+    ) -> Result<Self> {
+        let block = current_block::block_number(&block_stream.borrow())?;
+        Ok(Self {
+            mutexed: Mutex::new(Mutexed::new(
+                number_of_pairs_to_auto_update,
+                block,
+                maximum_recent_block_age,
+            )),
+            number_of_blocks_to_cache,
+            inner,
+            block_stream,
+        })
+    }
+
+    pub async fn update_cache(&self, new_block: u64) -> Result<()> {
+        let pairs = self
+            .mutexed
+            .lock()
+            .unwrap()
+            .recently_used_pairs()
+            .collect::<HashSet<_>>();
+        tracing::debug!("automatically updating {} pair pools", pairs.len());
+        let pools = self
+            .inner
+            .fetch(pairs.clone(), Block::Number(new_block))
+            .await?;
+        {
+            let mut mutexed = self.mutexed.lock().unwrap();
+            mutexed.insert(new_block, pairs.into_iter(), &pools);
+            let oldest_to_keep = new_block.saturating_sub(self.number_of_blocks_to_cache.get() - 1);
+            mutexed.remove_cached_blocks_older_than(oldest_to_keep);
+            mutexed.last_update_block = new_block;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait::async_trait]
+impl PoolFetching for PoolCache {
+    async fn fetch(&self, pairs: HashSet<TokenPair>, block: Block) -> Result<Vec<Pool>> {
+        let block = match block {
+            Block::Recent => None,
+            Block::Number(number) => Some(number),
+        };
+
+        let mut cache_hits = Vec::new();
+        let mut cache_misses = HashSet::new();
+        let last_update_block;
+        {
+            let mut mutexed = self.mutexed.lock().unwrap();
+            for &pair in &pairs {
+                match mutexed.get(pair, block) {
+                    Some(pools) => cache_hits.extend_from_slice(&pools),
+                    None => {
+                        cache_misses.insert(pair);
+                    }
+                }
+            }
+            last_update_block = mutexed.last_update_block;
+        }
+
+        if cache_misses.is_empty() {
+            return Ok(cache_hits);
+        }
+
+        let cache_miss_block = block.unwrap_or(last_update_block);
+        let uncached_pools = self
+            .inner
+            .fetch(cache_misses.clone(), Block::Number(cache_miss_block))
+            .await?;
+        {
+            let mut mutexed = self.mutexed.lock().unwrap();
+            mutexed.insert(cache_miss_block, cache_misses.into_iter(), &uncached_pools);
+        }
+
+        cache_hits.extend_from_slice(&uncached_pools);
```

Comment:
As a follow up, maybe we should record a prometheus metric for cache hits and cache misses?